src/util/graph.h

```diff
@@ -584,6 +602,92 @@ std::vector<typename N::node_indext> grapht<N>::get_reachable(
   return result;
 }
 
+/// Run recursive depth-limited search on the graph, starting
+/// from multiple source nodes, to find the nodes reachable within n steps.
+/// This function initialises the search.
+/// \param src The node to start the search from.
+/// \param forwards true (false) if the forward (backward) reachability
+/// should be performed.
+/// \param limit  limit on steps
+template <class N>
+std::vector<typename N::node_indext> grapht<N>::depth_limited_search(
+  const node_indext &src,
+  std::size_t &limit,
+  bool forwards) const
+{
+  std::vector<node_indext> start_vector(1, src);
+  return depth_limited_search(start_vector, limit, forwards);
+}
+
+/// Run recursive depth-limited search on the graph, starting
+/// from multiple source nodes, to find the nodes reachable within n steps.
+/// This function initialises the search.
+/// \param src The nodes to start the search from.
+/// \param forwards true (false) if the forward (backward) reachability
+/// should be performed.
+/// \param limit  limit on steps
+template <class N>
+std::vector<typename N::node_indext> grapht<N>::depth_limited_search(
+  const std::vector<node_indext> &src,
+  std::size_t &limit,
+  bool forwards) const
+{
+  std::vector<bool> visited(nodes.size(), false);
+
+  for(const auto &node : src)
+    visited[node] = true;
+
+  return depth_limited_search(src, limit, visited, forwards);
+}
+
+/// Run recursive depth-limited search on the graph, starting
+// from multiple source nodes, to find the nodes reachable within n steps
+/// \param src The nodes to start the search from.
+/// \param limit  limit on steps
+/// \param visited vector of booleans indicating whether a node has been visited
+/// \param forwards true (false) if the forward (backward) reachability
+/// should be performed.
+template <class N>
+std::vector<typename N::node_indext> grapht<N>::depth_limited_search(
+  const std::vector<node_indext> &src,
+  std::size_t &limit,
+  std::vector<bool> &visited,
```

Comment:
How about making this a `std::set<node_indext> &`, then while the membership test would be a little more expensive than a `vector<bool>` you could use it as the result, rather than copying the result after each recursive call. The function would look something like

```
depth_limited_search(node_indext from, std::size_t steps, std::set<node_indext> &result)
{
  for(const auto o : nodes[from].out)
    if(result.insert(o.first).second)
      depth_limited_search(o.first, steps - 1, result);
}
```